### PR TITLE
[SPARK-5134] Bump default hadoop.version to 2.4.0 in pom.xml.

### PR DESCRIPTION
**Pull request by ryan-williams**
Fixes intellij's inability to resolve many hadoop/yarn dependencies.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -119,7 +119,7 @@
     <mesos.classifier>shaded-protobuf</mesos.classifier>
     <slf4j.version>1.7.5</slf4j.version>
     <log4j.version>1.2.17</log4j.version>
-    <hadoop.version>1.0.4</hadoop.version>
+    <hadoop.version>2.4.0</hadoop.version>
```
Comment (srowen) [inline pom.xml]: Why would this help something related to IntelliJ? I use IJ and do not have any problems with the Maven build. I kind of want to bump the default too, even if it has some broader implications.

Comment (ryan-williams) [inline pom.xml]: filed this just before I finished composing [SPARK-5115](https://issues.apache.org/jira/browse/SPARK-5115) because I wanted to a PR ID to reference there :)

tl;dr is that IntelliJ doesn't resolve YARN imports because the spark's `yarn`-module POM inherits `hadoop.version` from the parent POM, and YARN didn't exist at `1.0.4`

Comment (ryan-williams) [inline pom.xml]: I'd guess that any `hadoop.version >= 2` would work for my/intellij's purposes, but I'll double-check that

**Comment by srowen** [comment]
OK, I don't think the issue in SPARK-5115 should be the motivation here, even if all else equal, I'd prefer to use Hadoop 2+ as a default. Really, for most purposes, the Hadoop version does not matter, and where it matters, you have to set this version to the desired version anyway.

**Comment by ryan-williams** [comment]
It sounds like you're confirming my suspicion that the default is not `1.0.4` for any good reason.

Meanwhile, [SPARK-5115](https://issues.apache.org/jira/browse/SPARK-5115) outlines a pretty basic developer-environment-sanity issue that `1.0.4` is causing.

If you want me to rewrite this PR to have a different stated motivation, let me know what I should say.

**Comment by ryan-williams** [comment]
changed this PR to target [SPARK-5134](https://issues.apache.org/jira/browse/SPARK-5134)

**Comment by srowen** [comment]
Hm. I'd like to merge this since Hadoop 2.x seems like a much more sensible default these days. 

This doesn't change whether Spark works with Hadoop 1.x and doesn't change what builds are released, etc. 

The thing I think it *will* change is what version of Hadoop the Maven artifacts appear to depend on. As we've discussed a lot, this shouldn't matter in theory. Still, a 2.x dependency strikes me as less likely to cause problems since people are much more on 2.x these days.

It will also mean a default "mvn compile" will show some warnings about deprecated Hadoop APIs, but hey.

Any other thoughts on this? I would like to get at least one more confirmation that this seems sensible before I push the button. 

@vanzin @pwendell

**Comment by vanzin** [comment]
The one negative side-effect I can see is that it will make it easier for people to inadvertently break the hadoop 1.x build. I know because I've done that (since all my build scripts hardcode the hadoop version at 2.something).

Other than that, I'd like to meet that guy who still uses hadoop 1, and talk him out of it. :-)

**Comment by srowen** [comment]
@vanzin I hear you, although the flip side is that now it's a little easier to do wrong by Hadoop 2.x. Of course, it is generally backwards compatible, but we all know it's not 100%. And this is what Jenkins is for too. And yes who is still using 1.x...

I'm inclined to make this change but will wait a bit more for other thoughts.

**Comment by srowen** [comment]
@pwendell I'd love to hear your thoughts one way or the other on this one. My feeling is we should do this, but want a third opinion.

**Comment by pwendell** [comment]
The purpose for this is to make intellij easier. But can't you just enable the hadoop profiles when important and get the same effects? Is it different?

I would have concerns with this (as I said on a few other occasions too) on the basis of changing people's dependency graphs substantially for downstream projects that link against our published pom's. There are embedded applications that use Spark and don't really care about being on newer Hadoop API's - for instance applications that primarily interact with S3, or they use their own storage. This gives them a major build change since Hadoop itself has a lot of transitive dependencies. 

I guess the path for such applications would be that they would exclude Spark's Hadoop and add their own Hadoop dependency with the older version?

**Comment by srowen** [comment]
Yeah I don't think IJ is the issue. Good point about the transitive deps. My concern is that at some point, the transitive deps of a really old Hadoop version cause more surprise and harm than the change does.

Those apps shouldn't be relying on transitive deps to begin with. They have to harmonize with the version they deploy on, and so should be fixing the version / profiles they build against even now. In an _ideal_ world, of course.

Is this something best reserved for, say, 1.4 or 2.x?

**Comment by srowen** [comment]
I think we can close this for now as it's probably not going to happen until 2.x. SPARK-5134 remains open.

**Comment by ryan-williams** [comment]
k, thanks for following up @srowen

**Comment by pwendell** [comment]
yeah - sorry to be a wet blanket. Basically my feeling is that we shouldn't optimize for our own convenience as developers in turn for making life more difficult (in a regressing way) for people building downstream projects based on Spark. We can make sure to clearly document how to import into intelliJ and help people enable profiles on import, etc.

**Comment by srowen** [comment]
@pwendell To be clear this didn't end up being anything to do with IntelliJ. It's a question of whether building vs Hadoop 1.0.4 or Hadoop 2.x by default is better. I think it reduces surprise both for developers and downstream consumers, but, at this stage it's a subtle publishing change even if in theory it doesn't affect people that use Spark artifacts "correctly". I think this should be updated for Spark 2.x at the latest.

**Comment by ryan-williams** [comment]
np @pwendell, I understand there's some complicated stability concerns here.

**Comment by pwendell** [comment]
Hey I commented on the JIRA, but some recent changes in the way we publish artifacts actually makes this more tenable of a change.

https://issues.apache.org/jira/browse/SPARK-5134
